apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/Numeric.cs

```diff
@@ -0,0 +1,663 @@
+﻿// Copyright 2022 Google LLC
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     https://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using System;
+using System.Collections.Concurrent;
+using System.Globalization;
+using System.Linq;
+using System.Numerics;
+using System.Text.RegularExpressions;
+
+namespace Google.Cloud.BigQuery.V2
+{
+    // Note: These types are internal and used for implementing BigQueryNumeric and
+    // BigQueryBigNumeric only, which are thoroughly tested.
+    // If we ever need to make these types public, then we need to test them on its own.
+    internal sealed class NumericType
+    {
+        private sealed class Parameters
+        {
+            internal Parameters(BigInteger scaledMinValue, BigInteger scaledMaxValue, int scale)
+            {
+                GaxPreconditions.CheckArgumentRange(scale, nameof(scale), 0, PowersOf10.Length - 1);
+                GaxPreconditions.CheckState(scaledMinValue <= scaledMaxValue,
```

Comment:
Presumably the efficiency of this bit doesn't really matter as it will only be called a couple of times.